rfcs/2022-02-24-11552-dd-agent-style-secret-management.md

```diff
@@ -0,0 +1,191 @@
+  be changed but the Datadog Agent uses `ENC[secret_key]`)
+
+Datadog Secret API, as per the official [doc][dd-secret-backend-exec-api]:
+
+- Communication happens on stdout/sdin
+- It uses plain text json
+
+Vector would run the user-provided executable and feed its standard input with the list of secrets to retrieve:
+
+```json
+{"version": "1.0", "secrets": ["secret1", "secret2"]}
+```
+
+The version used by the Datadog agent is 1.0. The user provided executable should then reply on its standard output with
+a JSON formatted string:
+
+```json
+{
+  "secret1": {"value": "secret_value", "error": null},
+  "secret2": {"value": null, "error": "could not fetch the secret"}
+}
+```
+
+**Note**: The version field can be used to introduce specific behaviour, one major useful thing that could be introduced
+is the ability for the user provided executable to provide a secrete expiration date. This is mentioned in the
+improvements section.
+
+New top level options to be added will be sitting inside the `secret` namespace, this would lead to something like:
+
+```toml
+[secret.local_exec]
+type = "exec"
+path = "/path/to/the/command"
+argument = "--config foo=bar"
+timeout = 5
+
+[secret.prod_vault]
+type = "vault"
+address = "https://vault.corp.tld/"
+token = "secret://local_exec/vault_token"
+timeout = 5
+
+[sources.system_logs]
+type = "file"
+includes = ["/var/log/system.log"]
+
+[sinks.app_logs]
+type = "datadog_logs"
+default_api_key = "secret://prod_vault/dd_api_2022_02"
+inputs = ["system_logs"]
+```
+
+The first implementation would only support the `exec` backend but with extensibility point clearly identified to easily
+implement additional backend if needed.
+
+Overall the behaviour for corner cases should follow what's in place for environmment variable interpolation as this is
+a very close feature. When a configuration reload happens, secrets shall also be refreshed.
+
+### Implementation
+
+A secret backend that would lie in `./src/config/secret.rs` and would call the user provided executable for secrets,
+cache secrets to avoid calling the backend for the same key multiple time. It would read the configuration file once to
+get its config before further processing, ideally env var interpolating should be supported (this should not be a
+problem). In `./src/config/builder.rs`, `load_builder_from_paths` will still be returning a complete configuration with
+placeholders replaced by secrets.
+
+The `ConfigBuilder` struct will get a new `secret` field (typed to something like
+`IndexMap<ComponentKey,SecretBackend>`), this means that `load_builder_from_paths` will then assert if this field is not
+empty  before returning to the caller, and if it, the config will be reloaded with this `SecretBackend` passed to
+downstream callee and hook the secret interpolation around the same point as [the environment variable
+interpolation][env-var-hook]:
+
+- All `SECRET[<backend>/<key>]` placeholders present in config will be collected.
```

Comment:
Good point this will be more consistent with everything else. I will switch to using a dot and explicitly state that backend name cannot contain a dot but that `SECRET[backend_xyz.a.secret.key]` will query `backend_xzy` for the secret key `a.secret.key`.